Converted notifications to card rows instead of table layout

CuriousLearner opened the pull request:
This is an attempt to fix Issue #1368 

@seanlip Here are some screenshots:

![screen shot 2016-03-07 at 21 46 36](https://cloud.githubusercontent.com/assets/8039608/13575529/62bdb3b0-e4af-11e5-961a-c766ea00b90c.png)
![screen shot 2016-03-07 at 21 46 47](https://cloud.githubusercontent.com/assets/8039608/13575531/63d73b04-e4af-11e5-8ad5-01ddf10f996c.png)
![screen shot 2016-03-07 at 21 46 57](https://cloud.githubusercontent.com/assets/8039608/13575534/6592fa3c-e4af-11e5-8584-624dfdca668d.png)

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2326,6 +2336,35 @@ div#ng-curtain {
+  float: right;
+}
+
+.clearfix:after {
```
This seems too general?

seanlip replied:
(Also, can you explain its purpose?)

CuriousLearner commented:
@seanlip this is for adjusting elements as per the wireframe, would you recommend something else?

CuriousLearner commented:
@seanlip Since each element is now in the div, so I'm floating it to ensure that it remains as per wireframe. Also clearfix ensure that the two divisions do not overlap.

seanlip commented:
Thanks for the explanation! We'll revisit clearfix later, I think -- there are bigger design issues to tackle first.

I've just looked at the mocks, and I think that, UI-wise, this doesn't look so good. Some notes:
- The individual 'things' really need some visual separation (e.g. borders).
- The username should be linked to the profile page of the user. (As an alternative, have you considered using the profile picture?)
- Is "(Feedback from a learner)" meant to be a message body or a subject line?
- It might be interesting to try having the top-left thing be "CL left a comment on 'Demo Exploration'" and to visually separate that from the message body (e.g. by bolding it).
- Perhaps the icon at the top left may be better as the profile picture of the user, or some sort of hybrid between that and the type of notification? I'm not sure.

More generally, though, I think this needs quite a bit of design work to make it actually look nice. At the moment it looks very untidy, especially on small screens. I'm not really a designer, though, so although I can give you some ideas for what not to do, I'm afraid I can't give you much advice for what to do (sorry!) -- you'll probably need to play around with different things, and have a keen eye open for what looks nice and conveys the information appropriately. I'd be happy to take another look when you've done another pass or two.

CuriousLearner commented:
@seanlip, alright, the notifications right now are distinguished once the user hovers over it. But I agree, that without hover there should be some sort of distinction between different notifications. I'll try to re-iterate and update here. Thanks for your feedback.

CuriousLearner commented:
>  The individual 'things' really need some visual separation (e.g. borders).

Implemented visual separation. Applied different backgrounds to alternate notification rows.

>   The username should be linked to the profile page of the user. (As an alternative, have you considered using the profile picture?)

But the whole notification is itself linked to exploration feedback. Should that be removed?

>   Is "(Feedback from a learner)" meant to be a message body or a subject line?

It is notification-subject as I can see from the CSS rules.

>   It might be interesting to try having the top-left thing be "CL left a comment on 'Demo Exploration'" and to visually separate that from the message body (e.g. by bolding it).

Implemented this, and now it looks a lot better.

>  Perhaps the icon at the top left may be better as the profile picture of the user, or some sort of hybrid between that and the type of notification? I'm not sure.

I'm not sure about it either. I guess it won't be good to have profile picture at this moment (till we actually have different avatars for each user)

Here I'm attaching screenshots:
![screen shot 2016-03-08 at 18 50 24](https://cloud.githubusercontent.com/assets/8039608/13602911/903f7610-e55f-11e5-896c-4c7861dba39a.png)
![screen shot 2016-03-08 at 18 50 55](https://cloud.githubusercontent.com/assets/8039608/13602912/91b871ae-e55f-11e5-8950-2962c0c1cb59.png)

Please suggest changes!

seanlip commented:
Thanks, this is looking a lot better! Having said that, can you try showing, say, five notifications? (At first I just thought there was a colour gradient from dark to light, I didn't realize they were meant to alternate.)

For the link: hovering on the username should cause it to have an underline, and clicking on it should go to the profile page -- I think we want this to become a standard pattern throughout the app. Clicking anywhere else on the notification (when there's no underline) should take the user to the notification destination.

Some more notes:
- Why are different icons being used for two "feedback" notifications? That's rather confusing.
- It's probably better (more informative) to have a snippet of the notification content instead of just its subject, but let's leave that to a future PR.
- The small screen display looks very awkward, IMO. I think it needs more work.
  - A possible idea: what if each notification was a 'card' on the grey background (so there's no white card which they lie on top of)? There would be no vertical spacing between the cards. Their backgrounds could all be white, but there are borders separating the different cards. (Just an idea -- it may or may not work.) This could probably apply to the large-screen version, too.

CuriousLearner commented:
Hi @seanlip, what do you think about this design?

![screen shot 2016-03-08 at 21 35 28](https://cloud.githubusercontent.com/assets/8039608/13607661/3a2451fc-e577-11e5-9720-2b0df59e1990.png)


Also, answering your previous queries:

> Why are different icons being used for two "feedback" notifications? That's rather confusing.
It was done prior to this PR. As I see in the code, these denotes 'exploration-commit' and 'feedback-threads'

> It's probably better (more informative) to have a snippet of the notification content instead of just its subject, but let's leave that to a future PR.

I couldn't understand what you meant by this.

PS: I'm still working on this and figuring out a way to make this efficient on small screens.

seanlip commented:
- I agree; I'd say "(Anonymous)" instead of leaving it blank.
- "exploration-commit" means that someone else has made a commit to an exploration the user manages. You seem to have called all of them "X left a comment on Y", which is incorrect in the "exploration-commit" case. The "exploration-commit" messages may need to be displayed differently, in fact; maybe they could be a different colour from the feedback-thread ones.
- Say a feedback message is left with subject "Subject" and text "This is some feedback". By "snippet" I meant show "This is some..." instead of "Subject".

CuriousLearner commented:
Hi @seanlip 

>    I agree; I'd say "(Anonymous)" instead of leaving it blank.

I would implement this.

>    "exploration-commit" means that someone else has made a commit to an exploration the user manages. You seem to have called all of them "X left a comment on Y", which is incorrect in the "exploration-commit" case. The "exploration-commit" messages may need to be displayed differently, in fact; maybe they could be a different colour from the feedback-thread ones.

Alright, I'll try a fix for this. IMHO, changing the color of specific message would disturb the whole UI, since rows are alternate in color and suddenly there would seem to be a color change out of no-where. But I would still try and show draft to you.

>    Say a feedback message is left with subject "Subject" and text "This is some feedback". By "snippet" I meant show "This is some..." instead of "Subject".

Okay, I'll implement this.

Would you take a look at the above image and share your opinion.

Also this is how it looks on small screens right now:
![screen shot 2016-03-08 at 22 17 16](https://cloud.githubusercontent.com/assets/8039608/13608610/89459c92-e57b-11e5-918a-4f194f6edb14.png)

CuriousLearner commented:
What would be the text in `exploration_commit` case?

seanlip commented:
> What should be the text in exploration_commit case?

I'd suggest: "X made a new commit to Demo Exploration"

> Alright, I'll try a fix for this. IMHO, changing the color of specific message would disturb the whole UI, since rows are alternate in color and suddenly there would seem to be a color change out of no-where. But I would still try and show draft to you.

Ah, I meant, don't alternate the row colours -- just colour-code the rows by the type of notification, or visually distinguish the two notification types in some way. I don't really care, as a user, about whether the rows are even or odd, and in any case there are other ways to visually distinguish different rows in a table.

Comments on small-screen mock:
- Still looks really messy (sorry). I don't know what you should do here -- perhaps get rid of the background 'card', make the individual tiles wider, shorten the text somehow, remove unnecessary information, redo the layout (e.g. that 'comment' icon on the top basically takes up a full line). I think I'd mainly suggest having a look at what other sites do, when they need to present similar information.

Comments on large-screen mock:
- I think you don't need the background card -- it's more that each individual notification is a "thing", rather than that the whole group of them is a "thing". Maybe each of the notification items can be individual cards.
- I still think the colour-coding is off, and it's also not really conveying information. It makes the whole thing feel pretty heavy IMO.
- Maybe we should use some other icon than a flag to represent a commit (I know this is a historical thing but it seems worth revisiting).
- Otherwise, the layout looks good to me.

These are mainly suggestions; it might also be worth getting some feedback from your friends or asking on the Gitter chat. Thanks!

CuriousLearner commented:
Hi @seanlip, 

I've tried to change the complete layout as per suggestions, removed elements that were not necessary and also adjusted the layout on small screens. Also, I've distinguished the commit notification by using another `backgound-color` and also changed the glyphicon.

The whole structure now treats every notification as card itself.
This is how it looks on both large and small screens:

![screen shot 2016-03-09 at 21 29 47](https://cloud.githubusercontent.com/assets/8039608/13641659/4b4f1304-e63f-11e5-9fe9-3a1f67b03700.png)

![screen shot 2016-03-09 at 21 30 01](https://cloud.githubusercontent.com/assets/8039608/13641658/49b8069a-e63f-11e5-8ee1-41a2f3d5a073.png)

Please let me know if any further changes are required.

seanlip commented:
A few quick comments:
* I think you want the each of the cards to be monochromatic and a different colour from the background (white, perhaps?) Also the green should extend to the entire card (otherwise the alignment is wonky) -- and perhaps you might choose a lighter colour; it's more of a background thing, after all.
* Also, try reducing the space between the cards, so that they are contiguous? I suspect that might be cleaner.
* The left borders of the cards should align perfectly with the left edge of the "Notifications" title.
* In the last card, on the mobile view, why has "Demo" dropped to a new line? That seems odd.

This might help, btw: http://www.dailyblogtips.com/crapthe-four-principles-of-sound-design/ (and look for other articles/posts discussing that principle).

CuriousLearner commented:
Hi @seanlip 

>   I think you want the each of the cards to be monochromatic and a different colour from the background (white, perhaps?) Also the green should extend to the entire card (otherwise the alignment is wonky) -- and perhaps you might choose a lighter colour; it's more of a background thing, after all.

I converted the background color to be white. However I'm unable to distinguish one particular notification type from another ( such that background of `md-card` is changed ). I've tried to it via JS, jQuery.parent() and even Angular selector. Why is it not changing is that the elements are created dynamically by Angular.

Also, I know there is event delegation in jQuery, and that might be the fix for this issue, but I'm confused about the event to use to trigger this thing. Since that event should be automatically triggered once all the DOM elements are loaded.

I've tried almost everything, but the real fix lies to trigger this:
```$('.oppia-exploration-commit-notification').parent().css('background-color', 'lightblue');```
once the DOM is loaded. If you can suggest something, that would be great :)

> Also, try reducing the space between the cards, so that they are contiguous? I suspect that might be cleaner.

Done.

>    The left borders of the cards should align perfectly with the left edge of the "Notifications" title.
>     In the last card, on the mobile view, why has "Demo" dropped to a new line? That seems odd.
> 

Done.

seanlip commented:
This is looking quite nice, IMO! Thanks @CuriousLearner. Some suggestions:
- maybe bold only the username, but nothing else? Otherwise it's a bit heavy.
- re the background colour -- I think it's fine for now, actually, if you leave everything white, but alternatively you can also use ng-class to apply conditional CSS classes based on some conditions. So something like ng-class="{'...': isFeedbackNotification(), '...': !isFeedbackNotification()}". Does that help?

CuriousLearner commented:
Hi @seanlip

Is it good now?

![screen shot 2016-03-11 at 19 35 46](https://cloud.githubusercontent.com/assets/8039608/13704168/8730c264-e7c0-11e5-8d3a-7443e780fc65.png)

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+        <md-card ng-repeat="notification in recentNotifications"
+            ng-class="{'oppia-dashboard-row-recent': lastSeenMsec && lastSeenMsec < notification.last_updated_ms && notification.author_username !== currentUsername, 'oppia-dashboard-row': !lastSeenMsec || lastSeenMsec >= notification.last_updated_ms || notification.author_username === currentUsername}"
+            ng-click="navigateToItem(notification.activity_id, notification.type)">
+            <div ng-if="notification.type === 'feedback_thread'" class="oppia-feedback-notification">
```
Deindent this by 2, since it's only one level more than the md-card on line 35. I think you have an extra closing div tag at the end, too.

CuriousLearner replied:
Done

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+                <span class="glyphicon glyphicon-comment"
+                      title="Feedback thread"></span>
+              </div>
+                <div>
```
Indentation wrong -- should be on same level as previous div

seanlip replied:
(Could you also check this sort of thing carefully, with regards to your changes? It's possible I might have missed other occurrences while reviewing.)

CuriousLearner replied:
Apologies for that. I'll be correcting all this and take care from next time positively.

CuriousLearner replied:
Done

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+            ng-click="navigateToItem(notification.activity_id, notification.type)">
+            <div ng-if="notification.type === 'feedback_thread'" class="oppia-feedback-notification">
+              <div>
+                <span class="label label-primary"
```
Why is this being repeated separately in both cases?

CuriousLearner replied:
Made it now in a single condition case.

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+                      title="Feedback thread"></span>
+              </div>
+                <div>
+                  <div class="oppia-notification-author-username"><strong><a ng-click="navigateToProfile($event, notification.author_username)"><[notification.author_username]></a><span ng-if="!notification.author_username">Anonymous</span></strong> left a comment on<span class="oppia-notification-activity-title"><[notification.activity_title]></span></div>
```
Drop the closing /div to the next line

CuriousLearner replied:
Done

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+                  <div class="oppia-notification-author-username"><strong><a ng-click="navigateToProfile($event, notification.author_username)"><[notification.author_username]></a><span ng-if="!notification.author_username">Anonymous</span></strong> left a comment on<span class="oppia-notification-activity-title"><[notification.activity_title]></span></div>
+                </div>
+              <div class="oppia-notification-last-updated"><[getLocaleAbbreviatedDatetimeString(notification.last_updated_ms)]></div>
+              <div class="clearfix"></div>
```
Do you really need the clearfix? I'd prefer to drop it if possible -- we've never needed to use it before, and this isn't the first time we're floating things in a div...

CuriousLearner replied:
IMHO clearfix is a better way for handling floating elements, and in this particular case, this would suffice. Also, if we remove this, we need to specify the height for the notification cards for multiple screen size. They won't scale with the flow of content.

If you think we should remove it then we'll remove it and figure another way around. Please let me know the needful.

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
-            <td><[notification.subject]></td>
-          </tr>
-        </table>
+            <div ng-if="notification.type === 'exploration_commit'" ng-init="changeExplorationCommitNotificationColor();" class="oppia-exploration-commit-notification">
```
No, don't use ng-init. What's wrong with putting it in the CSS class?

CuriousLearner replied:
Is this a bad practice?

seanlip replied:
In this case, I think so. The JS code should, in general, not reference the DOM.

CuriousLearner replied:
Not applied color change for different types of comment. As glyphicon were enough to distinguish them.

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,53 @@ <h2 class="oppia-page-heading">Notifications</h2>
+                      title="Exploration commit"></span>
+              </div>
+                <div ng-if="notification.type === 'exploration_commit'">
+                  <div class="oppia-notification-author-username"><strong><a ng-click="navigateToProfile($event, notification.author_username)"><[notification.author_username]></a></strong><span ng-if="!notification.author_username">Anonymous</span> made a new commit to<span class="oppia-notification-activity-title"><[notification.activity_title]></span></div>
```
Drop /div to next line.

But really, I don't like all this duplication. It's only the icon, the link, and a bit of the message that's changing, right? Better to have a single branch, and condition only for those three elements, rather than writing huge 'if' clauses (similar to the sort of thing that was done previously).

CuriousLearner replied:
Done

seanlip commented:
The UI is looking better, thanks @CuriousLearner! I made some comments on the code. I also have three UI questions:

(a) what does it look like on larger and smaller widths? I'm guessing larger widths look fine -- but, on smaller widths, how do you handle wrapping? (I actually think you should put the date on the first line on the right, have some space on the left of it, and either truncate or wrap the left bit. This implies a two-column layout, for which you could use the bootstrap grid or some other reasonable method.)

(b) what does the "New" label look like? From the code I saw, it seems like it might push the icon away. This seems a bit odd. You might consider doing away with that label altogether, and using a different colours to highlight the new tiles instead (maybe white for new tiles, light grey for seen tiles). I think it's fine if the "left a comment" and "made a new commit" tiles have the same colour by default -- the icon helps to distinguish them anyway.

(c) how does it look if you narrow the vertical space between the tiles even further, so that you get what looks like one light horizontal line between them? If it doesn't look nice, I think it's fine to leave it as it currently is.

CuriousLearner commented:
Hi @seanlip 

> (a) what does it look like on larger and smaller widths? I'm guessing larger widths look fine -- but, on smaller widths, how do you handle wrapping? (I actually think you should put the date on the first line on the right, have some space on the left of it, and either truncate or wrap the left bit. This implies a two-column layout, for which you could use the bootstrap grid or some other reasonable method.)

It looks like this right now:
![screen shot 2016-03-12 at 23 47 29](https://cloud.githubusercontent.com/assets/8039608/13724506/d0c632b8-e8ad-11e5-9144-77028de2ae1a.png)
![screen shot 2016-03-12 at 23 47 46](https://cloud.githubusercontent.com/assets/8039608/13724508/d26aeac8-e8ad-11e5-94b2-38eb38b942df.png)


Still need to use two column layout? 
> (b) what does the "New" label look like? From the code I saw, it seems like it might push the icon away. This seems a bit odd. You might consider doing away with that label altogether, and using a different colours to highlight the new tiles instead (maybe white for new tiles, light grey for seen tiles). I think it's fine if the "left a comment" and "made a new commit" tiles have the same colour by default -- the icon helps to distinguish them anyway.

Removed it all together, and the new notifications are seen as shown above.

> (c) how does it look if you narrow the vertical space between the tiles even further, so that you get what looks like one light horizontal line between them? If it doesn't look nice, I think it's fine to leave it as it currently is.

I couldn't get you on this. At how much width do you want to see this? Currently the mobile version is taken at 320 X 480.

seanlip commented:
Hi @CuriousLearner: 

(a) Looks nice (including mobile)! So, re two-column layout, maybe that's not needed after all.
(b) Great!
(c) This question is independent of the screen width -- I meant reducing the space _between_ the bottom border of one card and the top border of the next card. But, on second thoughts, don't worry about it; it looks fine for now.

One note: please address all code comments following the instructions in step 5 [here](https://github.com/oppia/oppia/blob/develop/CONTRIBUTING.md#instructions-for-making-a-code-change) -- this will help us keep track of what's been addressed and what hasn't. Let me know when you've done this, and I'll do another review pass. Thanks!

CuriousLearner commented:
HI @seanlip I've addressed the code comments, can you please check?

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,59 @@ <h2 class="oppia-page-heading">Notifications</h2>
+            <div ng-if="notification.type === 'exploration_commit'">
+              <div class="oppia-notification-author-username">
+                <strong>
+                  <a ng-click="navigateToProfile($event, notification.author_username)"><[notification.author_username]></a>
```
Should this have an ng-if="notification.author_username" on it?

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,59 @@ <h2 class="oppia-page-heading">Notifications</h2>
+            <div ng-if="notification.type === 'feedback_thread'">
+              <div class="oppia-notification-author-username">
+                <strong>
+                  <a ng-click="navigateToProfile($event, notification.author_username)"><[notification.author_username]></a>
```
Ditto -- ng-if?

Also, there's still significant duplication. I think you can reduce it.

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,59 @@ <h2 class="oppia-page-heading">Notifications</h2>
+            <div class="oppia-notification-last-updated">
+              <[getLocaleAbbreviatedDatetimeString(notification.last_updated_ms)]>
+            </div>
+            <div class="clearfix"></div>
```
Still not keen on this. What about creating two divs inside each card, one vertically above the other? In the first div, display everything except the date; in the second, display only the date (and float it right).

seanlip replied:
Ah -- actually, that's sort of what you did. What I mean is, for the second div, have oppia-notification-last-updated-container (which isn't floated right) and, within it, have oppia-notification-last-updated (which is floated right within oppia-notification-last-updated-container). Does that work?

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2322,10 +2335,46 @@ div#ng-curtain {
+
+.oppia-notification-author-username,
+.oppia-notification-subject {
+  float: left;
```
I'm actually a bit surprised that this is needed. Couldn't you just use a span element (to keep it inline) instead of a div?

seanlip replied:
(Ditto for oppia-notification-type)

seanlip commented:
Thanks, @CuriousLearner -- I've taken another review pass. Getting there!

CuriousLearner commented:
Hi @seanlip,

I tried to fix this issue, but at a certain widths, this is how the screen looks:

![screen shot 2016-03-15 at 00 07 09](https://cloud.githubusercontent.com/assets/8039608/13755699/721db542-ea42-11e5-93cb-4b811c817e88.png)

Tried various things to keep them up on card.

Rest for wider screens it looks good. 

Can you guide me how to go about this issue?

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2322,10 +2335,31 @@ div#ng-curtain {
+  float: right;
+}
+
+@media all and (max-width: 500px) {
```
Just checking: why 'all' here but not in line 117? (Not sure which is better, but the inconsistency seems odd.)

CuriousLearner replied:
I just did a quick search and all is never used in the entire code-base. So, I'm removing it for right now, and will search about it, and update here if I find something concrete.

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,47 @@ <h2 class="oppia-page-heading">Notifications</h2>
-              <span ng-if="notification.type === 'exploration_commit'"
-                    class="glyphicon glyphicon-flag"
-                    title="Exploration commit"></span>
+        <md-card ng-repeat="notification in recentNotifications"
```
If you give the md-card a style of "overflow: auto;", does that fix the floating problem?

CuriousLearner replied:
Yes, that fixes it! Thanks :)

seanlip commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -26,46 +26,47 @@ <h2 class="oppia-page-heading">Notifications</h2>
+        <md-card ng-repeat="notification in recentNotifications"
+          ng-class="{'oppia-dashboard-row-recent': lastSeenMsec && lastSeenMsec < notification.last_updated_ms && notification.author_username !== currentUsername, 'oppia-dashboard-row': !lastSeenMsec || lastSeenMsec >= notification.last_updated_ms || notification.author_username === currentUsername, '.oppia-exploration-commit-notification': lastSeenMsec && lastSeenMsec < notification.last_update ...
+          <div class="oppia-notification-content">
+            <span class="oppia-notification-type" >
```
nit: remove space before '>'

CuriousLearner replied:
Done

seanlip commented:
Hi @CuriousLearner -- I think I found a way to do it, and I also made two comments. Does it help?

CuriousLearner commented:
Hi @seanlip I addressed the review comments and did the changes. Please have a look. Meanwhile I'll search for media queries.

Thanks!

seanlip commented:
Just one more question! (Also, I think what you did with media queries is fine.)

CuriousLearner commented:
@seanlip Did the changes and it looks like this right now.

Works for recent notifications too. Have a look:

![screen shot 2016-03-15 at 18 10 22](https://cloud.githubusercontent.com/assets/8039608/13778081/c0ec3c6a-ead9-11e5-8588-27d9778eb602.png)

seanlip commented:
Looks fantastic. Thanks @CuriousLearner! Will merge when Travis checks pass.

CuriousLearner commented:
Thanks @seanlip :+1:
